tfjs-backend-webgpu/src/webgpu_program.ts

```diff
@@ -50,8 +50,285 @@ export interface WebGPUProgram {
+    }
+
+    fn ${funcName}Coords(coordsIn : ${type}) -> f32 {
+      var coords = coordsIn;
+      ${coordsSnippet}
+      return f32(${texName}[getIndexFromCoords${rankStr}(${
+      unpackedCoordsSnippet}, ${shapeStr})]);
+    }
+  `;
+}
+
+export function getMainHeaderAndGlobalIndexString(): string {
+  return `
+    ${getMainHeaderString()}
+      let index = getGlobalIndex();
+`;
+}
+
+export function getMainHeaderString(): string {
+  return `
+  ${getWorkGroupSizeString()}
+  fn main(@builtin(local_invocation_id) LocalId : vec3<u32>,
+          @builtin(global_invocation_id) GlobalId : vec3<u32>,
+          @builtin(num_workgroups) NumWorkgroups: vec3<u32>) {
+    localId = LocalId;
+    globalId = GlobalId;
+    numWorkgroups = NumWorkgroups;
+`;
+}
+
+/**
+ * Generates getOutputCoords() function that computes output coordinates from
+ * dispatch geometry to reduce arithmetic.
+ */
+export function getOutputCoordsSnippet(
+    outShape: number[],
+    dispatchLayout: {x: number[], y?: number[], z?: number[]}):
+    [string, number] {
+  const {x, y = [], z = []} = dispatchLayout;
+
+  const outRank = outShape.length;
+  if (x.length === outRank) {
+    const dtype = getCoordsDataType(outRank);
+    const snippet = `fn getOutputCoords() -> ${dtype}{
+      let globalIndex = getGlobalIndex();
+      return getCoordsFromIndex(globalIndex);
+    }
+    `;
+    return [snippet, outRank];
+  }
+
+  let gatherDimensionsStr = '';
+  const dims = [x, y, z];
+
+  let rank = 0;
+
+  for (let i = 0; i < dims.length; i++) {
+    const arr = dims[i];
+
+    if (arr.length === 0) {
+      continue;
+    }
+
+    rank += arr.length;
+
+    if (arr.length === 1) {
+      gatherDimensionsStr += `let d${arr[0]} = i32(globalId[${i}]);`;
+    } else {
+      const strides = symbolicallyComputeStrides(arr, 'uniforms.outShape');
+      gatherDimensionsStr += `var index${i} = i32(globalId[${i}]);`;
+      for (let j = 0; j < strides.length; j++) {
+        gatherDimensionsStr += `let d${arr[j]} = index${i} / ${strides[j]};`;
+
+        if (j === strides.length - 1) {
+          gatherDimensionsStr += `let d${arr[j + 1]} = ` +
+              `index${i} - d${arr[j]} * ${strides[j]};`;
+        } else {
+          gatherDimensionsStr +=
+              `index${i} = index${i} - d${arr[j]} * ${strides[j]};`;
+        }
+      }
+    }
+  }
+
+  const dimensions = [];
+  for (let i = 0; i < rank; i++) {
+    dimensions.push(`d${i}`);
+  }
+
+  const dtype = getCoordsDataType(rank);
+  let snippet = `fn getOutputCoords() -> ${dtype} {
+    ${gatherDimensionsStr}
+  `;
+  if (dimensions.length === 0) {
+    snippet += `return ${dtype}(0); }`;
+  } else {
+    snippet += `return ${dtype}(${dimensions.join(',')}); }`;
+  }
+
+  return [snippet, rank];
+}
+
+export function getWorkGroupSizeString(): string {
+  return `
+  @stage(compute) @workgroup_size(workGroupSizeX, workGroupSizeY, workGroupSizeZ)
+`;
 }
 
 export const makeBindGroup =
```

Comment:
After refactoring, this file is more like program and shader related. Maybe move `makeBindGroup` out of this file?

Reply:
Moved to backend_webgpu.ts